### PR TITLE
Add 'testing' build tag to test and testutil files

### LINKED ISSUE #310: Rename test related go files so no test specific files are exported
**Is your feature request related to a problem? Please describe.**

We should check all our test related go files (tests, test-related utils) and rename them with `_test.go` suffix (e.g., `testutil.go -> util_test.go`, `testutil_soci.go -> util_soci_test.go`). 

This ensures those test-specific helpers won't be accidentally imported into non-test packges. See https://github.com/awslabs/soci-snapshotter/pull/107#pullrequestreview-1141590789 for more context.

**Describe the solution you'd like**

Rename test specific util files ([for example](https://github.com/search?q=repo%3Aawslabs%2Fsoci-snapshotter+path%3A%22testutil.go%22&type=code)) so they won't be imported by accident.

### PR DESCRIPTION
**Pull request by rdpsin**
This commit adds the 'testing' build flag to all test and testutil files. Updated `make test` to use `-tags testing`. Running `go test` manually will now require passing `tags testing`.

Removed the function `getArtifactEntry` from `artifacts.go` since it was only used within a test.

Moved the struct `waiter` from `layer.go` to `testutils.go` since it was only used within a test.

Signed-off-by: Rishabh Singhvi <rdpsin@amazon.com>

*Issue #, if available:*
Closes #310 

*Description of changes:*

*Testing performed:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by kzys** [comment]
Is it hard to rename `testutil.go` to something like `util_test.go`, instead of adding `testing` tag?

I like Go project's consistency due to `gofmt` and its relatively rich standard packages. So I'm reluctant to introduce project-specific local rules like this.

**Comment by rdpsin** [comment]
We tried doing that initially. The problem is that quite a few of our tests rely on functionality that is present in other packages' `testutil.go`. For example, `metadata/db` tests rely on functionality present in `metadata/testutil.go`. Renaming them to `*_test.go` means that Go can't find the functions anymore.

**Comment by djdongjin** [comment]
> We tried doing that initially. The problem is that quite a few of our tests rely on functionality that is present in other packages' testutil.go.

If these utils are shared by other packages, should we just move them to our [`util/testutil`](https://github.com/awslabs/soci-snapshotter/tree/main/util/testutil) package? I feel it's common for projects to have a standalone `testutil` package ([kubernetes](https://github.com/kubernetes/kubernetes/tree/a58e8993922ca0cee85a9e982f66ea5c9039ec93/pkg/controller/testutil), [containerd](https://github.com/containerd/containerd/tree/3f565daf682f652a13bbafeff63bc618377e1b71/pkg/testutil), [nerdctl](https://github.com/containerd/nerdctl/tree/main/pkg/testutil)).

Personally I feel having a standalone `testutil` package is sufficient as a reminder that this code is for testing and avoids `accidentally be imported into non-test packges`.

If some utils are used only by the package's own test, then we can put them in the package's `util_test.go`. If later it's shared, we can move to `testutil`.

**Comment by rdpsin** [comment]
Yes, that makes sense. But the issue is that current `testutil` files use a lot of unexported types. Moving all `testutil`s to a separate package would be a much bigger change.

**Comment by djdongjin** [comment]
What if we first rename `testutil.go`/`testutils.go` to `util_test.go`? And if it's not possible (e.g. util shared by other package tests), then we move it to the existing `testutil` package. Since it's a package, we can use the previous package name as file name (e.g., `metadata/testutil.go` -> `testutil/metadata.go`)

**Comment by rdpsin** [comment]
I don't really want to do that. Like I mentioned, it would mean exporting all unexported fields. This would mean touching most (if not all) non-testing code as well.

File: soci/artifacts_test.go [hunk 1]
```diff
@@ -96,15 +98,6 @@ func TestGetIndexArtifactEntries(t *testing.T) {
 		t.Fatalf("the retrieved content should match to the original content")
 	}
 }
-
-func TestGetArtifactEntry_ArtifactDB_DoesNotExist(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (rdpsin) [reply soci/artifacts_test.go:100]: The function it was testing was only used within this test. So removed both the function and the test.

**Comment by kzys** [comment]
For `metadata` and `metadata/db`, if one needs another for testing, it may make sense to merge them together. db.NewReader() is only used from cmd/soci-snapshotter-grpc/main.go.

https://dave.cheney.net/practical-go/presentations/qcon-china.html#_consider_fewer_larger_packages

> The advice I find myself repeating is to prefer fewer, larger packages. Your default position should be to not create a new package. That will lead to too many types being made public creating a wide, shallow, API surface for your package..

Another small thing we could do is taking testing.TB as a parameter in test utilities. This would reminds developers that these functions are not intended for production.

**Pull request closed without merging**